### PR TITLE
- Split rank setup verification code and the binary [run-systemtest]

### PR DESCRIPTION
**Pull request by baldersheim**
- Accumulate errors during ranksetup. Report them at the end, and also verify them in some tests.

@havardpe PR

### CODE REVIEW

**Review by havardpe — COMMENTED** [review]
I suggest refactoring 'errors' into 'messages', where each message can be info/warning/error in order to preserve both severity and ordering of emitted log messages after code split.

File: searchcore/src/apps/verify_ranksetup/verify_ranksetup.cpp [hunk 1]
```diff
@@ -21,11 +21,12 @@
...
 #include <vespa/searchlib/fef/test/plugin/setup.h>
 #include <vespa/config/subscription/configsubscriber.hpp>
-#include <vespa/vespalib/util/signalhandler.h>
+#include <vespa/vespalib/util/stringfmt.h>
+#include <vespa/vespalib/stllike/asciistream.h>
 #include <optional>
 
 #include <vespa/log/log.h>
-LOG_SETUP("vespa-verify-ranksetup");
+LOG_SETUP(".verify-ranksetup");
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchcore/src/apps/verify_ranksetup/verify_ranksetup.cpp]: Removed

File: searchcore/src/apps/verify_ranksetup/verify_ranksetup.cpp [hunk 2]
```diff
@@ -183,31 +201,22 @@ App::verifyConfig(const VerifyRanksetupConfig &myCfg,
...
                            profile.fef.property[j].value);
         }
+        vespalib::string msg;
         if (verify(schema, properties, repo, expressions, models)) {
-            LOG(info, "rank profile '%s': pass", profile.name.c_str());
+            msg = fmt("rank profile '%s': pass", profile.name.c_str());
+            LOG(info, "%s", msg.c_str());
         } else {
-            LOG(error, "rank profile '%s': FAIL", profile.name.c_str());
+            LOG(error, "%s", msg.c_str());
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchcore/src/apps/verify_ranksetup/verify_ranksetup.cpp]: No logging here anymore

File: searchlib/src/vespa/searchlib/fef/verify_feature.cpp [hunk 3]
```diff
@@ -2,28 +2,29 @@
...
     BlueprintResolver resolver(factory, indexEnv);
     resolver.addSeed(featureName);
     bool result = resolver.compile();
     if (!result) {
-        LOG(error, "verification failed: %s (%s)",
-            BlueprintResolver::describe_feature(featureName).c_str(), desc.c_str());
+        const BlueprintResolver::Errors & compileErrors(resolver.getCompileErrors());
+        errors.insert(errors.end(), compileErrors.begin(), compileErrors.end());
+        vespalib::string msg = fmt("verification failed: %s (%s)",BlueprintResolver::describe_feature(featureName).c_str(), desc.c_str());
+        errors.emplace_back(msg);
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/fef/verify_feature.cpp]: Now there should be proper separation between warning and error.

**Comment by baldersheim** [comment]
@havardpe PTAL

File: searchlib/src/vespa/searchlib/fef/blueprintresolver.cpp [hunk 4]
```diff
@@ -311,27 +314,9 @@ BlueprintResolver::compile()
     executor.shutdown();
     size_t stack_usage = compiler.stack_usage();
     if (stack_usage > (128_Ki)) {
-        LOG(warning, "high stack usage: %zu bytes", stack_usage);
+        _compileErrors.emplace_back(fmt("high stack usage: %zu bytes", stack_usage));
```
[the remark this answers is not in the document]
Comment (havardpe) [reply searchlib/src/vespa/searchlib/fef/blueprintresolver.cpp:317]: I agree about it being of interest, I just do not like that _compileError and _compileErrors are inconsistent (compilation passes with compilation errors). I think we should rename _compileErrors to _compile(Warnings/Messages/???). Going for Messages tagged with severity all the way down would be the most robust, even if all messages are warnings.

**Pull request merged**